### ISSUE TITLE
📊 Expanded state on token balances

### ISSUE THREAD

**Issue by ricburton** [balance-io/balance-manager#214]
<img width="907" alt="screen shot 2018-05-14 at 19 51 41" src="https://user-images.githubusercontent.com/438160/40029106-af6001ac-57b0-11e8-9bf4-1ec6f93e8f81.png">

I have always thought it would be cool to see the value of all the tokens fluctuating over time like this. I'm imaging it being at the top of the balances page.

**Comment by ricburton** [comment]
I would love to integrate data from OnChainFX.com

![expanded balances](https://user-images.githubusercontent.com/438160/40050088-767f3ba6-5804-11e8-9c84-1001e16f9f08.png)

https://onchainfx.com/asset/ethereum

**Comment by ricburton** [comment]
I think a good V1 here is to add data from https://trivial.co

Here is the API: http://telegra.ph/API-draft-06-06

**Comment by ricburton** [comment]
I also think adding this data set in would be great: https://github.com/MyEtherWallet/ethereum-lists/blob/master/tokens/tokens-eth.json

```
{
		"symbol": "BLT",
		"address": "0x107c4504cd79C5d2696Ea0030a8dD4e92601B82e",
		"decimals": 18,
		"name": "Bloom",
		"ens_address": "",
		"website": "https://hellobloom.io",
		"logo": {
			"src": "https://etherscan.io/token/images/hellobloom_28.png",
			"width": 28,
			"height": 28,
			"ipfs_hash": ""
		},
		"support": {
			"email": "team@hellobloom.io",
			"url": "https://hellobloom.io/faq"
		},
		"social": {
			"blog": "https://blog.hellobloom.io",
			"chat": "",
			"facebook": "https://facebook.com/bloomtoken",
			"forum": "",
			"github": "https://github.com/hellobloom",
			"gitter": "",
			"instagram": "",
			"linkedin": "",
			"reddit": "https://www.reddit.com/r/BloomToken",
			"slack": "https://slack.hellobloom.io",
			"telegram": "https://t.me/joinchat/FFWDdQ1hxqIg3jLe7zBVFQ",
			"twitter": "https://twitter.com/bloomtoken",
			"youtube": "https://www.youtube.com/channel/UCaMfVoHQzwX47XKUawWVCzg"
		}
	},
```

**Comment by mfolkeseth** [comment]
Cryptocompare seems to have a neat and open API for getting news https://min-api.cryptocompare.com/

**Comment by mfolkeseth** [comment]
Trivial changed their endpoint today. New docs https://docs.google.com/document/d/1ytiXg2XmNbAqK12ne5-Of104qP55d3F-TpNs1K5Vyzg/edit

**Comment by ricburton** [comment]
Thanks @mfolkeseth. Excited to play with this.

**Comment by Polej** [comment]
Hi,

I'm from Trivial and we're thinking quite seriously about expanding token balance information with data from our API ( https://docs.google.com/document/d/1ytiXg2XmNbAqK12ne5-Of104qP55d3F-TpNs1K5Vyzg/edit ).

Firstly we would like to add the simplest token data, e.g. our token description.

Is it ok for me to (try to :) ) do it or maybe you have your own idea how to do it?

**Comment by mfolkeseth** [comment]
Hi @Polej ,

Thanks for reaching out!

I'm sorry, but I'm not quite sure what you mean. Can you elaborate? You are already have a lot if information in your api, including token description. 

EDIT: Ah, were you thinking about taking on this task?

**Comment by Polej** [comment]
I meant to modify the code of Balance Manager to load the data from the API. Both backend and frontend. But then I found out you already done it. Great! 👍 

Is the issue with setting proxy connected with CORS (Cross-Origin-Resource-Sharing), AJAX-ing the data from another host?

**Comment by mfolkeseth** [comment]
That is correct, or, we needed to create a proxy to solve the problem below.
`Failed to load https://trivial.co/api/tokeninformation?token_address=0xf230b790e05390fc8295f4d3f60332c93bed42e2: The 'Access-Control-Allow-Origin' header has a value 'http://localhost:8080' that is not equal to the supplied origin. Origin 'https://localhost:3000' is therefore not allowed access.`

**Comment by Polej** [comment]
I'll fix it

**Comment by mfolkeseth** [comment]
Thanks a lot!

By the way. I find your price chart and daily active users very neat. Any chance of exposing those data? And also. You also have a lot of other useful info on trivial.co like more than 2 news, discussions. Do you plan on expanding the public API in the future?

**Comment by Polej** [comment]
I think it should be fixed now. Not 100% sure, please test it :)

For this data I need to discuss with my team, I'll respond tomorrow. I think it is highly possible to share this data :)

**Comment by mfolkeseth** [comment]
Like a charm :) Thanks @Polej!
You can follow the update on #269

**Comment by ricburton** [comment]
Can you post screenshots? I’m on mobile and eager to see it 🤗

**Comment by mfolkeseth** [comment]
![screen shot 2018-06-13 at 08 26 07](https://user-images.githubusercontent.com/7047431/41333850-8f233040-6ee3-11e8-9eb3-3216a347b02a.png)

**Comment by mfolkeseth** [comment]
Try one of your tokens at: https://deploy-preview-269--balance-manager.netlify.com/metamask

![screen shot 2018-06-13 at 08 27 58](https://user-images.githubusercontent.com/7047431/41333894-bfc01e7a-6ee3-11e8-8e79-5272d5791953.png)

**Comment by Polej** [comment]
@mfolkeseth 
> By the way. I find your price chart and daily active users very neat. Any chance of exposing those data? And also. You also have a lot of other useful info on trivial.co like more than 2 news, discussions. Do you plan on expanding the public API in the future?

Yes we may expand API in the nearest future, probably even this week.
It's nice to see first use in such short time.

What data do you need?
1. Price chart data 
2. Daily active users chart data
3. More news (how much?)
4. Discussions (how much?)
Maybe anything else?

I'm trying to group all the requests in one place and find out what to add in the next iteration.

**Comment by mfolkeseth** [comment]
Maybe you can weigh in here @ricburton ? If we are going to have discussions and more news, it would be useful to have pagination.

**Comment by ricburton** [comment]
This is a great start! Today is a launch day so we're busy with that. Will think on all this more.

Thanks for all your hard work :)

Richard

Sent via Superhuman ( https://sprh.mn/?vip=richard@balance.io )

**Comment by mfolkeseth** [comment]
@Polej, by the way. Any chance this API can return similar information about ethereum itself?

**Comment by Polej** [comment]
@mfolkeseth We are currently working on ethereum data.

**Comment by Polej** [comment]
At this moment our API has been a bit updated (to give results for ETH), but we don't collect much data about it yet.

Ethereum data is accessible here: (token_address=eth or ETH)
https://trivial.co/api/tokeninformation?token_address=eth

Many of the API fields are nulled now, we will gradually add them.

The reason to post this small API change is to give you, @mfolkeseth a possibility to update the Balance Manager build. Maybe the crash was because of lacking ETH data? Hope it will somehow solve the problem.

**Comment by mfolkeseth** [comment]
Really great stuff! I will try to make some changes to the build over the weekend to include ETH. 

The cause of the crash was due in a proxy which was not included in the build. Will fix that also. I’ll let you know once I make the changes. Thanks!

**Comment by mfolkeseth** [comment]
@ricburton how about using the same toggle function as transactions? I like it!

**Comment by Polej** [comment]
Hello, a little update:
**Trivial.co** now shares **price chart** and **daily active users chart data** for tokens (as seen on our token page, for example here https://trivial.co/t/0xf230b790e05390fc8295f4d3f60332c93bed42e2 )

**Price chart data** is accessible at **/api/token-price-chart** with GET parameter **token_address** (address of token smart contract on Ethereum network):
Example:
https://trivial.co/api/token-price-chart?token_address=0xf230b790e05390fc8295f4d3f60332c93bed42e2

**Daily active users chart data** is accessible at **/api/daily-active-chart** with GET parameter **token_address**:
Example:
https://trivial.co/api/daily-active-chart?token_address=0xf230b790e05390fc8295f4d3f60332c93bed42e2

Please let me know if there are any issues :)

**Comment by ricburton** [comment]
Damn this is cool.

Cannot wait to integrate this more deeply.

Richard

Sent via Superhuman ( https://sprh.mn/?vip=richard@balance.io )

**Comment by mfolkeseth** [comment]
A little later than anticipated but the new integration is here #311.

Will update daily active users and price charts to the pull request soon. Real nice work @Polej!

**Comment by ricburton** [comment]
Out right now but stoked to see this.

Any screenshots? :)

**Comment by mfolkeseth** [comment]
The build went trough on the PR, so the preview is up if you wanna look around. Here is a screenshot. My plan is to add the price charts on to the right.


<img width="872" alt="screen shot 2018-06-19 at 00 43 42" src="https://user-images.githubusercontent.com/7047431/41577389-dd6e9548-738c-11e8-97a1-38644b4baf6b.png">

**Comment by Polej** [comment]
Amazing, looks brilliant @mfolkeseth ! <3

**Comment by ricburton** [comment]
This is fantastic! Thank you so much to both of you.

Cannot wait to play around with this :)

**Comment by mfolkeseth** [comment]
Added some graphs for active users and price via recharts.

Some feedback @Polej. I see that the price charts endpoint returns empty array for tokens that have no data, whereas `eth` gets null value. I think it would be beneficial to return the same value for the same result to avoid unexpected behaviour and unnecessary checks.

Also, would it be possible to have price charts and active users as includes for the token info endpoint? That way we don't have to make multiple request. 

All in all, the API is damn good to work with and responds really fast! 🎊 

***Big screen***
<img width="967" alt="screen shot 2018-06-19 at 12 44 04" src="https://user-images.githubusercontent.com/7047431/41593588-dea88854-73c0-11e8-8c1b-9a1ae6c5fa06.png">

***Smaller screen***
<img width="570" alt="screen shot 2018-06-19 at 12 44 24" src="https://user-images.githubusercontent.com/7047431/41593600-e77aa61a-73c0-11e8-87ce-a88803aaee1e.png">

**Comment by ricburton** [comment]
This is absolutely incredible! Amazing work. ❤️

We need a warning for TRON though. 🤓

**Comment by Polej** [comment]
@mfolkeseth Thanks for the work! And thanks for the feedback. I'll take your hints into consideration, not sure how quickly though :) Putting it on high priority.

About price chart/DAU chart data inside **/tokeninformation/** :
Current design is motivated by the fact that not everyone using API will need all the data. However your argument is very important, doing everything in one request may be very helpful.

Maybe a compromise, a Golden mean solution will suit our needs:
Passing additional parameter(s) like **price_chart_data**=true and/or **dau_chart_data**=true will fill additional fields.

What do you think about such solution?

**Comment by mfolkeseth** [comment]
@Polej yes, my thoughts exactly. I just didn't explain it very well 😃 Let's go with that 👍 
 
I totally agree with your initial design about not sending too data over the wire unnecessarily. That's an approach I personally embrace, in all API's, especially in our case where the client requests the API directly. 

Thanks again for being so on to this. It has made the integration way smoother and pleasant.

**Comment by Polej** [comment]
Ok, now there are 3 new GET parameters:

**price_chart_data** - setting it to 1 or True or true gives non-null field 'price_by_day' with price chart data (just like in /api/token-price-chart/), when default the field 'price_by_day' is null

**dau_chart_data** - setting it to 1 or True or true gives non-null field 'daily_active_by_day' with daily active users (DAU) chart data (just like in /api/daily-active-chart/), when default the field 'daily_active_by_day' is null

**news_count** - you can set how many news items are given, default is 4, if you don't need any news set it to 0, max is 10 (setting more gives at most 10)

Example:
https://trivial.co/api/token-information?token_address=0xb5a5f22694352c15b00323844ad545abb2b11028&dau_chart_data=1&price_chart_data=True&news_count=15

Little tweak of api route: now you should use **/api/token-information**, with hyphen **-**, but old version will still work.

**Comment by ricburton** [comment]
Thanks @Polej. This is going to be great. 

Are you working on highlighting the disastrous projects like EOS? We are happy to do that.

**Comment by ricburton** [comment]
This thread is amazing: https://twitter.com/blockchainjoy/status/1009136079096541184

Some highlights:


![](https://pbs.twimg.com/media/DgEssoxVQAAYQH5.jpg)

![](https://pbs.twimg.com/media/DgEu_I0VMAAJHUo.jpg)

**Comment by Polej** [comment]
@ricburton We don’t plan to add this at the moment, however we’re planning implementation of commenting box, where people can add comments like this one, and this will come alive pretty soon.

And to both of you @ricburton and @mfolkeseth , would it be possible to place a link to our token site "For more information visit Trivial.co", just like this one?

<img width="965" alt="screen shot 2018-06-20 at 11 40 46" src="https://user-images.githubusercontent.com/2440412/41651715-7ef9444a-7481-11e8-929a-2bebc819a2a1.png">

**Comment by ricburton** [comment]
I would love to expose the data source. I think this is great.

Richard

Sent via Superhuman ( https://sprh.mn/?vip=richard@balance.io )

**Comment by mfolkeseth** [comment]
@Polej of cource :) what is the address of that image? Maybe it’s better for you to have it on your side, rather than that every consumer app implements their own? If you want them to look similar I mean. Although it’s totally up to you 👍

**Comment by mfolkeseth** [comment]
@ricburton I’m keen to hear your thoughts or goal on how sophisticated the token info graphs should be.

**Comment by ricburton** [comment]
@mfolkeseth I will keep thinking on this. I am not 100% sure at the moment. Just love seeing this different data. So handy.

@christianbaroni any thoughts here?

**Comment by Polej** [comment]
@mfolkeseth 
The image button path is https://trivial.co/static/for_more_info.png
It should link to https://trivial.co/t/<token_ethereum_address>
API gives this URL in a field "token_page", or you can construct it from "address" (token eth address)

The image will probably (slightly) change, we need to work on it a bit more.

**Comment by mfolkeseth** [comment]
@Polej alright, so I've updated the api route to `/api/token-information`. No need to keep it deprecated, for us at least 😄 I have also replaced three requests to use just one, with include parameters instead.

As for the more visual changes, I've also added the data source link to trivial.co.

You can view the last build here: https://deploy-preview-311--balance-manager.netlify.com

***Big screen***
![screen shot 2018-06-22 at 16 52 54](https://user-images.githubusercontent.com/7047431/41783908-5e98a1b4-763e-11e8-8be3-6709bc450658.png)

***Small screen***
![screen shot 2018-06-22 at 16 53 34](https://user-images.githubusercontent.com/7047431/41783921-68090284-763e-11e8-8a85-1416bd463fa2.png)

**Comment by Polej** [comment]
@mfolkeseth Very good! Our need of linking to our site is satisfied :) Also good to hear about API use changes. 

Amazing work!

**Comment by ricburton** [comment]
This is awesome! Amazing work. 

@jinchung is back on Monday <3

Richard

Sent via Superhuman ( https://sprh.mn/?vip=richard@balance.io )

**Comment by Polej** [comment]
Hey @mfolkeseth, I've discovered bug when clicking on ETH, it looks the same as before: black screen.

@ricburton How much possible is deploying these features to master (after fixing this bug)? :)

**Comment by mfolkeseth** [comment]
@Polej thanks for the report. I'll look into it this evening. Cheers.

**Comment by mfolkeseth** [comment]
@Polej Turns out the reason why Ethereum crashed the app in this particular case was that there was no check whether the news object had source or not. Turns out some of them are `null` :) So I added a check for that and the app is working just fine. 

One question tho; what's up with the date formats? I saw that the graphs and news have separate date formats (number for graphs and some other string format for news). In both cases `date-fns@alpha-11` has trouble parsing those. This may be `date-fns`'s problem, but either way, is it possible to look into this and use a common format that works for all platforms?

Other changes:
- Added some margins to the graphs as some numbers were eaten by the borders
- Replaced `moment` with `date-fns`, because `date-fns`uses es6 modules and thus enables tree shaking. Smaller bundles 🎉

**Comment by Polej** [comment]
Hey, our Trivial API routes have been updated. The syntax of API URL is now starting with:
> https://api.trivial.co/

Detailed doc for current API is [here](https://docs.google.com/document/d/1ytiXg2XmNbAqK12ne5-Of104qP55d3F-TpNs1K5Vyzg/edit?usp=sharing)

@mfolkeseth Is this issue still a trouble?

We had much to do regarding our new website: [https://trivial.co/](https://trivial.co/) (Check it!)
Now there's a bit more space for fine-tuning our API.

**Comment by mfolkeseth** [comment]
Hey @Polej! 

I guess that's up to @ricburton to decide whether or not this issue deserves some attention. I'd be happy to update the PR with the last API changes.

Love the new site! 🎉
